test/e2e/activator_test.go

```diff
@@ -0,0 +1,175 @@
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = make(map[string]string)
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations["autoscaling.knative.dev/maxScale"] = "10"
+	}
+
+	test.CleanupOnInterrupt(func() { TearDown(clients, names, logger) }, logger)
+	defer TearDown(clients, names, logger)
+
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(logger, clients, &names, &configOptions, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", domain, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	responseChannel := make(chan *spoof.Response, concurrency)
+
+	sendRequests(client, url, concurrency, responseChannel, timeout, logger, t)
+
+}
+
+func sendRequests(client *spoof.SpoofingClient, url string, concurrency int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) {
+	t.Helper()
+	var (
+		group        errgroup.Group
+		responses    int32
+		wantResponse = http.StatusOK
+	)
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+
+	// Send out the requests asynchronously and wait for them to finish.
+	logger.Info("Starting to send out the requests")
+
+	// Print out stats.
+	go func() {
+		for {
+			select {
+			case <-time.Tick(5 * time.Second):
+				logger.Infof("Received responses: %d", atomic.LoadInt32(&responses))
+			case <-timeoutChan:
+				return
+			}
+		}
+	}()
+
+	// Send requests async and wait for the responses.
+	for i := 0; i < concurrency; i++ {
+		group.Go(func() error {
+			req, err := http.NewRequest(http.MethodGet, url, nil)
+			if err != nil {
+				return fmt.Errorf("error creating http request: %v", err)
+			}
+			res, err := client.Do(req)
+			if err != nil {
+				return fmt.Errorf("unexpected error sending a request, %v\n", err)
+			}
+			atomic.AddInt32(&responses, 1)
+			resChannel <- res
+			return nil
+		})
+	}
+	if err := group.Wait(); err != nil {
+		errChan <- fmt.Errorf("unexpected error making requests against activator: %v", err)
+	}
+	logger.Info("Done sending out requests")
+	close(resChannel)
+
+	logger.Info("Waiting for all requests to finish")
+
+	select {
+	case resp, ok := <-resChannel:
```

Comment:
Should this be in a loop?

Reply:
I don't think so, since we dispatch the messages once all requests were returned, it seems they are processed in a batch.

Reply:
This will still need to be a loop though. The current code will only handle the very first response.

Reply:
Given you've broken open the goroutine that calls `Wait` I think we can ditch the select here completely (and the errChan with it).

You can then read from the channel using:

```golang
for resp := range resChannel {
  ...
}
```

It will automatically stop once the channel is empty and closed.

Reply:
Yeap, we need the for indeed. 

Regarding the example with `for ...`, we could ditch the `errChan` and exit with FatalF directly, but we won't know what happened with other requests. I think it is better to read the errors from both errorChan and resChan.